### ISSUE TITLE
Group Norm Error When using FP16

### ISSUE THREAD

**Issue by yu45020** [pytorch/pytorch#17216]
## To Reproduce
```
import torch
x = torch.randn(1, 10, 20, 20).cuda().half()
M1 = torch.nn.BatchNorm2d(10).cuda()
M2 = torch.nn.GroupNorm(10, 10).cuda()
M1(x)  (tensor)
M2(x)  (error)
```
```
RuntimeError: Expected object of scalar type Float but got scalar type Half for argument #3 'tensor1'
  ```

## Expected behavior
Group norm seems to reshape the input tensor and then use batch norm internally based on [these lines]:(https://github.com/pytorch/pytorch/blob/7157be86226b79e3a698cbb0c524fc3ae59e28de/aten/src/ATen/native/Normalization.cpp#L465)

```
# ATen/native/Normalization.cpp

auto input_reshaped = input.contiguous().view({1, b * num_groups, -1});

    auto out = at::batch_norm(input_reshaped, {}, {}, {}, {}, true, 0, eps,
                              cudnn_enabled);
out = out.view(input_shape);
```
I am able to train models that have batch norms in mix precision with Nvidia's [apex](https://github.com/NVIDIA/apex), so I guess it should work for group morns. 

## Environment

PyTorch version: 1.0.1
Is debug build: No
CUDA used to build PyTorch: 10.0
OS: Microsoft Windows 10 Home
GCC version: (GCC) 7.3.0
CMake version: version 3.12.18081601-MSVC_2
Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130

**Comment by fmassa** [comment]
The error probably happens in https://github.com/pytorch/pytorch/blob/7157be86226b79e3a698cbb0c524fc3ae59e28de/aten/src/ATen/native/Normalization.cpp#L478-L485

Also worth noting that if we disable cudnn (via `torch.backends.cudnn.enabled = False`), then the BatchNorm2d case fails as well.

**Comment by yu45020** [comment]
> 
> 
> The error probably happens in
> 
> [pytorch/aten/src/ATen/native/Normalization.cpp](https://github.com/pytorch/pytorch/blob/7157be86226b79e3a698cbb0c524fc3ae59e28de/aten/src/ATen/native/Normalization.cpp#L478-L485)
> 
> Lines 478 to 485 in [7157be8](/pytorch/pytorch/commit/7157be86226b79e3a698cbb0c524fc3ae59e28de)
>      if (weight.defined() && bias.defined()) { 
>        return bias.view(affine_param_shape).addcmul(out, weight.view(affine_param_shape), 1); 
>      } else if (weight.defined()) { 
>        return out.mul(weight.view(affine_param_shape)); 
>      } else { 
>        return out.add(bias.view(affine_param_shape)); 
>      } 
>  } 
> 
> Also worth noting that if we disable cudnn (via `torch.backends.cudnn.enabled = False`), then the BatchNorm2d case fails as well.

I agree with you. I turn the group norm into batch norm-like function in PyTorch. fp 32 weights are multiplied with fp 16 inputs, which raises an error. 

I am reading papers in mix precision training. Group norm doesn't need to update moving mean/var, so I guess we can use it in fp 16. I run several experiments and find fp16 group norm only decreases around 3%  validation accuracy.

**Comment by alexbrowngh** [comment]
I can confirm that @fmassa is correct. The problem is that the "out" is fp16 and other things are fp32. Are there any solutions to this?

**Comment by gchanan** [comment]
We should check if this is still an issue after type promotion.

**Comment by mruberry** [comment]
This is expected and not fixed by our current type promotion. You can put your modules in .half() to fix the issues. Amp will automatically convert parts of your network to fp16 and convert the inputs to these portions of your network to half, too, then back to float32 when going through the fp32 parts of your network. 

PyTorch now has its own native AMP implementation provided by NVIDIA, by the way. See https://pytorch.org/docs/master/notes/amp_examples.html.
